arduino-ide-extension/src/browser/contributions/sketchbook.ts

```diff
@@ -1,28 +1,112 @@
+        const timeout = 5_000;
+        const results = await Promise.allSettled([
+          waitForEvent(this.settingsService.onDidChange, timeout),
+          waitForEvent(this.configService.onDidChangeSketchDirUri, timeout),
+        ]);
+        if (results.every(({ status }) => status === 'fulfilled')) {
+          this.doUpdate(true);
+        } else {
+          console.warn(
+            'Expected the sketchbook location to change but it did not.'
+          );
+        }
+      }
+    });
+    const maybeUpdateCurrentSketchbookPath = async (
+      sketchDirUri: URI | undefined = this.configService.tryGetSketchDirUri()
+    ) => {
+      if (sketchDirUri) {
+        const candidateSketchbookPath = await this.fileService.fsPath(
+          sketchDirUri
+        );
+        if (candidateSketchbookPath !== this._currentSketchbookPath) {
+          this._currentSketchbookPath = candidateSketchbookPath;
+        }
+      }
+    };
+    this.configService.onDidChangeSketchDirUri(
+      maybeUpdateCurrentSketchbookPath
+    );
+    maybeUpdateCurrentSketchbookPath();
   }
 
   override async onReady(): Promise<void> {
-    this.update();
+    this.windowService
+      .isFirstWindow()
+      // only the first window should warn about invalid sketch folder names.
+      .then((firstWindow) => this.doUpdate(firstWindow));
   }
 
   protected override update(): void {
-    this.sketchService.getSketches({}).then((container) => {
-      this.register(container);
-      this.menuManager.update();
-    });
+    // on regular menu updates, do not raise warning dialogs about invalid sketch folder names
+    this.doUpdate(false);
+  }
+
+  private doUpdate(raiseInvalidSketchFoldersWarning?: boolean): void {
+    this.sketchService
+      .getSketches({})
+      .then(({ container, sketchesInInvalidFolder }) => {
+        if (
+          raiseInvalidSketchFoldersWarning &&
+          sketchesInInvalidFolder.length
+        ) {
+          Promise.all(
+            sketchesInInvalidFolder.map(async (ref) => {
+              const fsPath = await this.fileService.fsPath(new URI(ref.uri));
+              return {
+                name: ref.name,
+                path: fsPath,
+              };
+            })
+          ).then((dialogInputs) =>
+            dialogInputs.reduce(async (acc, { name, path }) => {
+              await acc;
+              return remote.dialog.showMessageBox(remote.getCurrentWindow(), {
+                title: nls.localize(
+                  'arduino/sketch/ignoringSketchWithBadNameTitle',
+                  'Ignoring sketch with bad name'
+                ),
+                message: nls.localize(
+                  'arduino/sketch/ignoringSketchWithBadNameMessage',
+                  'The sketch "{0}" cannot be used. Sketch names must contain only basic letters and numbers. (ASCII-only with no spaces, and it cannot start with a number). To get rid of this message, remove the sketch from {1}',
```

Comment:
```suggestion
                  'The sketch "{0}" cannot be used. Sketch names must start with a letter or number, followed by letters, numbers, dashes, dots and underscores. Maximum length is 63 characters. To get rid of this message, remove the sketch from {1}',
```

The information re: "cannot start with a number" is obsolete: https://github.com/arduino/Arduino/commit/4545283ae773b6351d089e37b6d4080905ea8fd5

I used the exact text already stored in `arduino/sketch/invalidSketchName` as a replacement so perhaps that same object can be reused here to avoid duplication.